lxd/instance/drivers/driver_common.go

```diff
@@ -804,11 +806,86 @@ func (d *common) snapshotCommon(inst instance.Instance, name string, expiry *tim
 		return err
 	}
 
+	// Snapshot root disk.
 	err = pool.CreateInstanceSnapshot(snap, inst, d.op)
 	if err != nil {
 		return fmt.Errorf("Create instance snapshot: %w", err)
 	}
 
+	// Snapshot attached disk volumes (if requested).
+	if diskVolumesMode == api.DiskVolumesModeAllExclusive {
+		// Freeze instance for crash-consistent snapshots.
+		if inst.IsRunning() && !inst.IsFrozen() {
```

Comment:
This is another reason why I think we need to take the instance freezing logic out of `pool.CreateInstanceSnapshot` because this is not going to cause a consistent snapshot to be taken.

Consider the scenario when `diskVolumesMode == api.DiskVolumesModeAllExclusive`:

1. Call `pool.CreateInstanceSnapshot` - this freezes the instance, snapshots the root volume and unfreezes the instance. So now the instance is running again, writing to disks etc, diverging from snapshot we just took.
2. Then the instance is frozen again and the custom volume snapshots are taken, but they are not consistent with the root volume snapshot taken in 1. because the instance was unfrozen.